### DISCUSSION TITLE
Replica out of sync due to Needle creation on replication errors

### DISCUSSION THREAD

**Discussion by kmlebedev** [seaweedfs/seaweedfs#3275] [General]
@chrislusf Is it possible to correctly roll back the creation of the needle if the file could not be written to the replica?

1. First we write the file locally to needle 
https://github.com/chrislusf/seaweedfs/blob/master/weed/server/volume_server_handlers_write.go#L49
2. Next, write in replicas
https://github.com/chrislusf/seaweedfs/blob/master/weed/server/volume_server_handlers_write.go#L56
3. And if an error occurs, then the local record remains, which is why we get a different number of files in the replicas

**Comment by chrislusf** [comment]
the orphaned entries are deleted in filer layer:
 https://github.com/chrislusf/seaweedfs/blob/master/weed/server/filer_server_handlers_write_upload.go#L130

**Reply by ingardm** [reply]
What happens to the failed replication if not using filer?

**Reply by kmlebedev** [reply]
> the orphaned entries are deleted in filer layer: https://github.com/chrislusf/seaweedfs/blob/master/weed/server/filer_server_handlers_write_upload.go#L130

@chrislusf Lost records are not actually deleted, but are marked as deleted?

**Reply by chrislusf** [reply]
why it is "not actually deleted"?

**Reply by chrislusf** [reply]
> What happens to the failed replication if not using filer?

We should add something similar to filer to backtrack the written entries.

**Comment by kmlebedev** [comment]
@chrislusf  Why not add something like this?
```
	ret := operation.UploadResult{}
	isUnchanged, writeError := topology.ReplicatedWrite(vs.GetMaster, vs.grpcDialOption, vs.store, volumeId, reqNeedle, r)

	if writeError != nil {
		httpStatus = http.StatusInternalServerError
		ret.Error = writeError.Error()
		if _, err := vs.store.DeleteVolumeNeedle(volumeId, reqNeedle); err != nil {
			ret.Error += " DeleteVolumeNeedle: " + err.Error()
		}
	}
```

**Comment by kmlebedev** [comment]
For context, I get missed entries at volume.check.disk quite often. And I'm trying to understand where desynchronization can occur
